logstash-core/src/main/java/org/logstash/common/clients/HttpClient.java

```diff
@@ -0,0 +1,655 @@
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.http.Header;
+import org.apache.http.StatusLine;
+import org.apache.http.client.methods.*;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.message.BasicHeader;
+
+import javax.net.ssl.*;
+
+/**
+ * Easy-to-use HTTP client.
+ */
+public class HttpClient {
+    public enum Protocol { HTTP, HTTPS }
+
+    private final CloseableHttpClient httpClient;
+    private final URL baseUrl;
+
+    private HttpClient(CloseableHttpClient httpClient, URL baseUrl) {
+        this.httpClient = httpClient;
+        this.baseUrl = baseUrl;
+    }
+
+    /**
+     * Performs an HTTP HEAD request
+     *
+     * @param relativePath  Relative path to resource, e.g. api/status
+     * @throws RequestFailedException
+     * @throws IOException
+     */
+    public void head(String relativePath) throws RequestFailedException, IOException {
+        head(relativePath, null);
+    }
+
+    /**
+     * Performs an HTTP HEAD request
+     *
+     * @param relativePath  Relative path to resource, e.g. api/status
+     * @param headers       Headers to include with request
+     * @throws RequestFailedException
+     * @throws IOException
+     */
+    public void head(String relativePath, Map<String, String> headers) throws RequestFailedException, IOException {
+        String url = makeUrlFrom(relativePath);
+        HttpHead request = new HttpHead(url);
+
+        if (headers != null) {
+            headers.forEach(request::addHeader);
+        }
+
+        try (CloseableHttpResponse response = httpClient.execute(request)) {
+            if (response.getStatusLine().getStatusCode() >= 400) {
+                throw new RequestFailedException("HEAD", url , response.getStatusLine().getReasonPhrase());
+            }
+        } catch (IOException e) {
+            throw new RequestFailedException("HEAD", url ,e);
+        }
+    }
+
+    /**
+     * Performs an HTTP GET request
+     *
+     * @param relativePath  Relative path to resource, e.g. api/kibana/dashboards/export
+     * @return Response body
+     * @throws RequestFailedException
+     */
+    public String get(String relativePath) throws RequestFailedException, IOException {
+        return get(relativePath, null);
+    }
+
+    /**
+     * Performs an HTTP GET request
+     *
+     * @param relativePath  Relative path to resource, e.g. api/kibana/dashboards/export
+     * @param headers       Headers to include with request
+     * @return Response body
+     * @throws RequestFailedException
+     * @throws IOException
+     */
+    public String get(String relativePath, Map<String, String> headers) throws RequestFailedException, IOException {
```

Comment:
I like that we are mostly using `String` for bodies. 

One question, should we consider returning a wrapped version of the entity instead? Otherwise, I do foresee us adding `getBytes` in the future.

So, you would call `httpClient.get(...).toString()`, which would be a little more verbose, but maybe cleaner in the future than having to duplicate all the methods into a different format?

Curious what you think as well @original-brownbear .

Reply:
Yeah, I've been on the fence about this one as well (that returning a `String` might be too constraining). So thanks for nudging me over. I'm +1 to returning the `ByteArrayOutputStream` object instead.

Reply:
Chatted with @andrewvc off-PR and we agreed that, for now, this method should return an object of a custom class (one that we create). For now that class will have just one method, `toString()`. In the future we can add more methods to it as we need them for retrieving the content type, content length, content encoding, raw bytes, etc.